interim_results.py

```diff
@@ -0,0 +1,128 @@
+            stream.close()
+
+        async def sender(ws):
+            nonlocal clock_cursor, audio_cursor, transcript_cursor
+            clock_start = time.perf_counter()
+            try:
+                while True:
+                    data = await audio_queue.get()
+                    clock_cursor = time.perf_counter() - clock_start
+                    audio_cursor += float(len(data)) / float(RATE) / 2.0
+                    await ws.send(data)
+            except Exception as e:
+                print(f'Error while sending: {str(e)}')
+                raise
+
+        async def receiver(ws):
+            nonlocal clock_cursor, audio_cursor, transcript_cursor
+            transcript = ''
+            last_word_end = 0.0
+            latest_final_result_processed = False
+            async for msg in ws:
+                msg = json.loads(msg)
+
+                transcript_cursor = msg['start'] + msg['duration']
+
+                if msg['is_final'] and len(msg['channel']['alternatives'][0]['transcript']) > 0:
+                    latest_final_result_processed = False
+                    transcript += ' '
+                    transcript += msg['channel']['alternatives'][0]['transcript']
+                    last_word_end = msg['channel']['alternatives'][0]['words'][-1]['end']
+                    line = f'Clock: {clock_cursor:07.3f}, Audio: {audio_cursor:07.3f}, Transcript: {transcript_cursor:07.3f}; {transcript}'
+                    print(line)
+                    # using end='\r' doesn't work if the line wraps
+                    # this moves the cursor up and overwrites the line regardless of length
+                    # https://stackoverflow.com/a/47170056
+                    print("\033[{}A".format(len(line) // int(terminal_size.columns) + 1), end='')
+
+                elif not msg['is_final']:
+                    if len(msg['channel']['alternatives'][0]['transcript']) > 0:
+                        current_word_begin = msg['channel']['alternatives'][0]['words'][0]['start']
+                        if current_word_begin - last_word_end > SILENCE_INTERVAL:
+                            if len(transcript) > 0:
+                                print(f'Clock: {clock_cursor:07.3f}, Audio: {audio_cursor:07.3f}, Transcript: {transcript_cursor:07.3f}; {transcript}')
+                                beepy.beep(sound=1)
+                                transcript = ''
+                            latest_final_result_processed = True
+                    else:
+                        if transcript_cursor - last_word_end > SILENCE_INTERVAL:
+                            if len(transcript) > 0:
+                                print(f'Clock: {clock_cursor:07.3f}, Audio: {audio_cursor:07.3f}, Transcript: {transcript_cursor:07.3f}; {transcript}')
+                                beepy.beep(sound=1)
+                                transcript = ''
+                            latest_final_result_processed = True
```

Comment:
this algorithm only gets the case where the trigger happens from an interim result following a final result - it doesn't demonstrate the case where the trigger ought to happen in a final result (there are at least 2 edge cases there)

Reply:
I think this (simplified algorithm) is more correct:

```
import json

def find_index(messages):
    last_word_end = 0.0
    latest_final_result_processed = False
    
    index = 0
    for message in messages:
        message = json.loads(message)
        
        transcript_cursor = message['start'] + message['duration']

        # if there are any words in the message
        if len(message['words']) > 0:
            # if there is more than 1 word in the message, look to see if there is a gap of 2 seconds between any of the words
            if len(message['words']) > 1:
                previous_word_end = message['words'][0]['end']
                should_break = False
                for word in message['words']:
                    current_word_start = word['start']
                    if current_word_start - previous_word_end >= 2.0:
                        should_break = True
                        break
                    previous_word_end = word['end']
                if should_break:
                    break

            # if the last word in a previous message is 2 or more seconds older than the first word in this message
            current_word_begin = message['words'][0]['start']
            if current_word_begin - last_word_end >= 2.0 and last_word_end != 0:
                break
            last_word_end = message['words'][-1]['end']
        # if there were no words in this message, check if the the last word in a previous message is 2 or more seconds older than the timestamp at the end of this message
        elif transcript_cursor - last_word_end >= 2.0 and last_word_end != 0:
                break
                        
        index += 1

    return index
```

Reply:
Updated PR with this algorithm